### PR TITLE
Make the round tripper use the dialer's context

### PR DESCRIPTION
**Pull request by deweerdt**
this is useful in getting https://pkg.go.dev/net/http/httptrace events related to DNS when using the round tripper to make a request. Other events related to the connection continue not to work.

### CODE REVIEW

**Comment by deweerdt** [comment]
this addresses the DNS part of https://github.com/quic-go/quic-go/issues/3342

File: http3/roundtrip.go [hunk 1]
```diff
@@ -278,10 +278,35 @@ func isNotToken(r rune) bool {
...
+func resolveUDPAddrWithCtx(ctx context.Context, netstr string, addr string) (*net.UDPAddr, error) {
+	var (
+		err        error
+		host, port string
+		portnum    int
+	)
+	if host, port, err = net.SplitHostPort(addr); err != nil {
+		return nil, err
+	}
+	if portnum, err = net.DefaultResolver.LookupPort(ctx, netstr, port); err != nil {
```
Comment (marten-seemann) [inline http3/roundtrip.go:290]: Why is this needed? Why not just use the `port`? Does it cause a potential DNS roundtrip?

File: http3/roundtrip.go [hunk 2]
```diff
@@ -278,10 +278,35 @@ func isNotToken(r rune) bool {
...
+		return nil, err
+	}
+	if host == "" {
+		return &net.UDPAddr{IP: net.IPAddr{}.IP, Port: portnum, Zone: net.IPAddr{}.Zone}, nil
+	}
+	ips, err := net.DefaultResolver.LookupIPAddr(ctx, host)
```
Comment (marten-seemann) [inline http3/roundtrip.go:296]: I imagine that this will soon invite a request to use a different resolver here. How shall we deal with that?

**Review by marten-seemann — COMMENTED** [review]
Is there any precedent for moving address resolution into the HTTP layer? Does HTTP/2 do that?

**Comment by marten-seemann** [comment]
@deweerdt Are you still planning to work on this PR?

**Comment by marten-seemann** [comment]
Closing due to inactivity. @deweerdt Feel free to reopen once you have the bandwidth to push this through.
